compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,263 @@
+      if (auto vectorOperand = dyn_cast<VectorValue>(operand)) {
+        operand = getDistributed(rewriter, vectorOperand, options);
+      }
+      operands.push_back(operand);
+    }
+
+    // Get the new distributed vector types for the operation.
+    SmallVector<Type> resultTypes;
+    for (Value result : op->getResults()) {
+      Type resultType = result.getType();
+
+      // Distribute vector result types.
+      if (auto vectorResult = dyn_cast<VectorValue>(result)) {
+        resultType = VectorType::get(options.getDistributedShape(vectorResult),
+                                     vectorResult.getType().getElementType());
+      }
+      resultTypes.push_back(resultType);
+    }
+
+    // Replace the original op with the distributed op.
+    Operation *distributedOp =
+        rewriter.create(op->getLoc(), op->getName().getIdentifier(), operands,
+                        resultTypes, op->getAttrs());
+    replaceOpWithDistributedValues(rewriter, op, options,
+                                   distributedOp->getResults());
+    return success();
+  }
+
+private:
+  VectorLayoutAnalysis &analysis;
+  VectorLayoutOptions &options;
+};
+
+/// A rewriter for the pattern rewriting driver.
+class VectorDistributionRewriter : public PatternRewriter,
+                                   RewriterBase::Listener {
+public:
+  VectorDistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps)
+      : PatternRewriter(context, this), erasedOps(erasedOps) {}
+
+  /// We need to keep track of erased operations so that we never try to
+  /// distribute them again.
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+private:
+  // Reference to operations to be erased and the worklist of the driver.
+  DenseSet<Operation *> &erasedOps;
+};
+
+static bool canDistribute(Operation *op, VectorLayoutAnalysis &analysis) {
+  auto values = llvm::to_vector_of<Value>(op->getOperands());
+  llvm::append_range(values, op->getResults());
+
+  // Check if all operands and results of this operation have a layout.
+  return llvm::all_of(values, [&](Value value) -> bool {
+    if (auto vectorValue = dyn_cast<VectorValue>(value)) {
+      return analysis.getLayout<Attribute>(vectorValue) != nullptr;
+    }
+    return false;
+  });
+}
+
+static void
+debugPrintUniqueOperationNames(SmallVectorImpl<Operation *> &worklist,
+                               DenseSet<Operation *> &exclude) {
+  DenseSet<StringRef> uniqueNames;
+  for (Operation *op : worklist) {
+    if (exclude.contains(op))
+      continue;
+    uniqueNames.insert(op->getName().getStringRef());
+  }
+
+  for (StringRef name : uniqueNames) {
+    llvm::dbgs().indent(2) << "* " << name << "\n";
+  }
+  LLVM_DEBUG(llvm::dbgs() << "\n");
+}
+
+static void applyVectorDistributionDriver(
+    Operation *root, const FrozenRewritePatternSet &patterns,
+    VectorLayoutAnalysis &analysis, VectorLayoutOptions &options) {
+
+  DenseSet<Operation *> erasedOps;
```

Comment:
Are erased ops deleted? If so, don't we have to store `uintptr_t` to avoid using pointers to deleted objects which is UB?

Reply:
In llvm there are 2 methods: `eraseFromParent` (remove and delete) and `removeFromParent` (remove only), so to me `erasedOps` looks like ops that were deleted which seems off on its own.

Reply:
OK I found this code in PatternMatch.cpp:
```c++
    rewriteListener->notifyOperationRemoved(op);

    // Explicitly drop all uses in case the op is in a graph region.
    op->dropAllUses();
    op->erase();
  };
```

So if we remove the op pointer immediately in the callback it's fine, if we still store pointers after the callback and use later, even in comparisons, it's UB.

Reply:
Not using this approach anymore so resolved.